src/dotty/tools/dotc/typer/ProtoTypes.scala

```diff
@@ -430,6 +430,31 @@ object ProtoTypes {
       (if (theMap != null) theMap else new WildApproxMap).mapOver(tp)
   }
 
+  trait AssigningProto extends ProtoType
+
```

Comment:
I wonder whether you need the trait and the case class. Instead of all this, maybe you can do it as with AnyFunctionProto:

    @sharable object AssignProto extends UncachedGroundType with MatchAlways

Can you try this and see whether it would work?